File: Affine.js
class Rect {
    #key_order;
    constructor(key) {
        if (key.match(/[^A-Za-z]/)) {
            throw new Error("Key must only contain alphabet characters.");
        }
        if (key.split("").sort().join("").match(/(.)\1/)) {
            throw new Error("Key must avoid repeat characters.");
        }
        if (key.match(/[a-z]/)) {
            key = key.toUpperCase();
        }
        this.key = key;
        this.n = key.length;
        this.#key_order = [...key].sort().map(x => [...key].indexOf(x));
    }

    #encode(text) {
        return text.toUpperCase().replace(/[^A-Z]/g, "");
    }

    set_key(new_key) {
        this.__init__(new_key);
    }

    #pad(text, n) {
        let need_pad = -text.length % n;
        let pad = "";
        for (let i = 0; i < need_pad; i++) {
            pad += String.fromCharCode(Math.floor(Math.random() * 26) + 65);
        }
        return text + pad;
    }

    encrypt(text) {
        if (typeof text !== "string") {
            throw new TypeError("Input should be string");
        }
        text = this.#encode(text);
        text = this.#pad(text, this.n);
        let cipher = [];
        let text_list = [...text];
        for (let i = 0; i < text_list.length; i += this.n) {
            cipher.push(...this.#key_order.map(x => text_list[i + x]));
        }
        return [cipher.join(""), this.#key_order];
    }

    decrypt(text, key) {
        if (typeof text !== "string") {
            throw new TypeError("Input should be string");
        }
        if (text.length % this.n !== 0) {
            throw new Error(`The length of input should be a mutiple of ${this.n}`);
        }
        if (text.match(/[^A-Z]/)) {
            throw new Error("Invalid ciphertext: input contains non-alphabet characters");
        }
        let text_list = [...text];
        let plain_text = [];
        for (let i = 0; i < text_list.length; i += this.n) {
            plain_text.push(...this.#key_order.map(x => text_list[i + x]));
        }
        return plain_text.join("");
    }
}

File: Affine Cipher/Affine.py
from re import sub
sub_encrpty = {"A":0,"B":1,"C":2,"D":3,"E":4,"F":5,"G":6,
       "H":7,"I":8,"J":9,"K":10,"L":11,"M":12,
       "N":13,"O":14,"P":15,"Q":16,"R":17,"S":18,
       "T":19,"U":20,"V":21,"W":22,"X":23,"Y":24,
       "Z":25}
sub_decrypt = {0:"A",1:"B",2:"C",3:"D",4:"E",5:"F",6:"G",
       7:"H",8:"I",9:"J",10:"K",11:"L",12:"M",
       13:"N",14:"O",15:"P",16:"Q",17:"R",18:"S",
       19:"T",20:"U",21:"V",22:"W",23:"X",24:"Y",
       25:"Z"}
class Affine:
    def __init__(self,a:int,b:int) -> None:
        self.a = a
        self.b = b
        self.__inverse_a = self.__xgcd(a,26)[0]
    def __xgcd(self,a,b):
        if b ==0:
            return 1,0
        else:
            x,y = self.__xgcd(b,a%b)
            x,y = y,(x-(a//b)*y)
            return x,y
    def show_key(self):
        return self.a,self.b

    def __encode(self,text: str):
        stripped = sub(r"[^a-zA-Z]", "", text)
        return stripped.upper()

    def set_key(self,a:int,b:int):
        self.__init__(a,b)


    def encrypt(self,text:str):
        text = self.__encode(text)
        cipher = "".join([sub_decrypt[(self.a*sub_encrpty[letter]+self.b)%26] for letter in text])
        return cipher

    
    def decrypt(self,text:str):
        if text.isalpha()==False:
            raise Exception("Input should contain only alphbets")
        if text.isupper() == False:
            raise Exception("Input should be upper case letters")
        plain_text = "".join([sub_decrypt[self.__inverse_a*(sub_encrpty[letter]-self.b)%26] for letter in text])
        return plain_text
a = Affine(a=7,b=17)
d = a.encrypt("VYOHVUTWVHPZUWTZUUBEYDUB")
print(a.decrypt(d))


    
        


